### PR TITLE
Laboratory work #1, Kakoulina Tasya - 23FPL2

### PR DESCRIPTION
**Pull request by tasya-kakoulina**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -5,9 +5,22 @@
 """
 # pylint:disable=too-many-locals, unused-argument, unused-variable
 
+# откуда берется language(3) + округлить мсе + closest profile:/
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:8]: Remove

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -5,9 +5,22 @@
...
+
 
 def tokenize(text: str) -> list[str] | None:
-    """
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:12]: revert the docstring

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -5,9 +5,22 @@
...
+        text = text.lower
+        tokenized_list = []
+        for i in text():
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:15]: Amazing move, please try to find more usual way to call `lower`

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/main.py:12]: use early return - https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -19,11 +32,19 @@ def tokenize(text: str) -> list[str] | None:
         list[str] | None: A list of lower-cased tokens without punctuation
 
     In case of corrupt input arguments, None is returned
-    """
+"""
 
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
-    """
+    if isinstance(tokens, list) and all(isinstance(i, str) for i in tokens):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:39]: again - early return

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -33,11 +54,17 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
 
     In case of corrupt input arguments, None is returned
-    """
+"""
 
 
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
-    """
+    if isinstance(language, str) and isinstance(text, str):
+        lang_profile_dict = {'name': language, 'freq': calculate_frequencies(tokenize(text))}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:62]: you can return dict "as is", without creating the variable

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -48,11 +75,21 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
         dict[str, str | dict[str, float]] | None: A dictionary with two keys – name, freq
 
     In case of corrupt input arguments, None is returned
-    """
+ """
 
 
 def calculate_mse(predicted: list, actual: list) -> float | None:
-    """
+    if isinstance(predicted, list) and isinstance(actual, list) and len(predicted) == len(actual):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:82]: early return

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Good job! Return docstring and try for 10!

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -20,6 +22,14 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(text, str):
+        return None
+    tokenized_list = []
+    for i in text.lower():
+        if i.isalpha():
+            tokenized_list += i
+    en_text = tokenized_list
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:31]: why do you create the `en_text` variable?

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -76,14 +107,36 @@ def compare_profiles(
     Args:
         unknown_profile (dict[str, str | dict[str, float]]): A dictionary of an unknown profile
         profile_to_compare (dict[str, str | dict[str, float]]): A dictionary of a profile
-            to compare the unknown profile to
+        to compare the unknown profile to
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:110]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -93,18 +146,29 @@ def detect_language(
 ) -> str | None:
     """
     Detect the language of an unknown profile.
-
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:96]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -93,18 +146,29 @@ def detect_language(
...
     Args:
         unknown_profile (dict[str, str | dict[str, float]]): A dictionary of a profile
             to determine the language of
         profile_1 (dict[str, str | dict[str, float]]): A dictionary of a known profile
         profile_2 (dict[str, str | dict[str, float]]): A dictionary of a known profile
-
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:102]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -93,18 +146,29 @@ def detect_language(
...
     Returns:
         str | None: A language
-
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:105]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -119,6 +183,13 @@ def load_profile(path_to_file: str) -> dict | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(path_to_file, str):
+        return None
+    with open(path_to_file, 'r', encoding='utf-8') as file_to_profile:
+        profile = json.load(file_to_profile)
+    if isinstance(profile, dict):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:190]: better - `if not ` -> `return None`

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -130,11 +201,30 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
 
     Returns:
         dict[str, str | dict] | None: A dict with a lower-cased loaded profile
-            with relative frequencies without unnecessary n-grams
+        with relative frequencies without unnecessary n-grams
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:204]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -167,6 +267,15 @@ def detect_language_advanced(
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(unknown_profile, dict) or not isinstance(known_profiles, list):
+        return None
+    profiles_list = []
+    for i in known_profiles:
+        profiles_list.append((i['name'], compare_profiles(unknown_profile, i)))
+    profiles_list.sort(key=lambda x: (x[-1], x[0]))
+    if profiles_list:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:276]: Why do you add this check after sort?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Great, I like it!

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py:105]: done

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py:283]: done

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -105,6 +158,20 @@ def detect_language(
...
+    mse_2 = compare_profiles(unknown_profile, profile_2)
+    if mse_1 is None or mse_2 is None:
+        return None
+    if not isinstance(profile_1['name'], str) or not isinstance(profile_2['name'], str):
+        return None
+    if mse_1 < mse_2:
+        return profile_1['name']
+    if mse_1 > mse_2:
+        return profile_2['name']
+    return None
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:174]: if equal?

File: lab_1_classify_by_unigrams/main.py [hunk 16]
```diff
@@ -135,6 +209,25 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+    if not isinstance(profile, dict) or not all(k in profile for k in ('freq', 'name', 'n_words')):
+        return None
+    processed_profile = {'name': profile['name'],
+                         'freq': {}}
+    d = {}
+    for i in profile['freq']:
+        if isinstance(i, str) and len(i) == 1:
+            d[i] = profile['freq'][i]
+    letters_list = list(d)
+    for j in letters_list:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:221]: `j`, `d`... bad names! I understood nothing.

File: lab_1_classify_by_unigrams/main.py [hunk 17]
```diff
@@ -178,3 +295,8 @@ def print_report(detections: list[tuple[str, float]]) -> None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(detections, tuple):
+        return None
+    for i in detections:
+        print(f'{i[0]}: MSE {i[-1]:.5f}')
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:301]: just print? Why?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 17 above)
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py:299]: done (? i hope)

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -3,7 +3,9 @@
 
 Language detection
 """
+import copy
```
[the remark this answers is not in the document]
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 19]
```diff
@@ -34,6 +43,12 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(tokens, list) or not all(isinstance(token, str) for token in tokens):
+        return None
+    frequency_dict = {}
+    for token in tokens:
+        frequency_dict[token] = tokens.count(token) / len(tokens)
```
[the remark this answers is not in the document]
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 20]
```diff
@@ -135,6 +208,25 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+    for token in profile['freq']:
+        if isinstance(token, str) and len(token) == 1:
+            freq_dict[token] = profile['freq'][token]
+    letters_list = list(freq_dict)
+    for letter in letters_list:
+        if not letter.isupper():
+            freq_dict[letter] = freq_dict.get(letter, 0) + freq_dict.get(letter.upper(), 0)
+            processed_profile['freq'][letter] = freq_dict[letter] / profile['n_words'][0]
+            if letter.upper() in freq_dict:
+                del freq_dict[letter.upper()]
```
[the remark this answers is not in the document]
Comment (tasya-kakoulina) [reply lab_1_classify_by_unigrams/main.py]: done

**Review by demid5111 — APPROVED** [review]

**Pull request merged**